data-prep-pipeline/scripts/subtype_tree/create_subtype_tree.py

```diff
@@ -0,0 +1,618 @@
+
+            - oncotree (pandas df)
+
+        Outputs:
+            - table node (pd.Series): one row of the new oncotable, which includes 
+                each parent of said node
+        '''
+
+        #Oncotree indexes starting at 1, we want to index starting at 0    
+        cur_level = node.NodeLevel
+        levels = dict({
+                    'OncotreeCode':node.OncotreeCode,
+                    'NodeName':node.NodeName,
+                    'NodeLevel':node.NodeLevel-1, #index at 0
+                    f'Level{cur_level-1}':node.OncotreeCode #index at 0
+                })
+
+        while node.parent != 'TISSUE':
+            node = oncotree.query('OncotreeCode == @node.parent').iloc[0]
+            cur_level = node.NodeLevel
+            levels[f"Level{cur_level-1}"] = node['OncotreeCode']
+        
+        return pd.Series(levels)
+
+    oncotable = oncotree[oncotree.OncotreeCode != 'TISSUE']\
+                    .apply(
+                        find_all_parents,
+                        **{'oncotree':oncotree},
+                        axis=1
+                    ).assign(
+                        DepmapModelType = lambda x: x.OncotreeCode,
+                        NodeSource = "Oncotree"
+                    )
+    
+    #relabel duplicated node names
+    dup_idx = oncotable.loc[oncotable.NodeName.duplicated(keep=False)].index
+
+    oncotable.loc[dup_idx, 'NodeName'] = oncotable.loc[dup_idx].apply(
+        lambda x: f"{x.NodeName} ({x.Level0})",
+        axis=1
+    )
+    
+    return oncotable
+
+def construct_new_table_node(new_code,
+                            new_name,
+                            new_source,
+                            parent_code,
+                            oncotable,
+                            code_col='DepmapModelType'):
+    '''
+    A function to create a new custom node in the subtype tree
+
+    Inputs:
+        - new_code (str): the code to use for the new node
+        - new_name (str): the name to use for the new node
+        - new_source (str): the data source of the new node (e.g. Depmap)
+        - parent_code (str): the code of the new node's parent
+        - oncotable (pandas df): the table-version of oncotree
+        - code_col (str): which code column to use to find the parent and add
+            the new node's code
+
+    Outputs:
+        - new_node (pd.Series): a new custom node with all of the necessary information 
+    '''
+
+    #find the node of the parent
+    parent_node = oncotable[oncotable[code_col] == parent_code].iloc[0]
+
+    #copy the parent, and add all of the new information
+    new_node = parent_node.copy()
+
+    #reset all node meta fields
+    new_node['NodeLevel'] = parent_node['NodeLevel'] + 1
+    new_node['NodeName'] = new_name
+    new_node['NodeSource'] = new_source
+    new_node[code_col] = new_code
+    new_node['OncotreeCode'] = np.nan #if we are adding a custom node, there is no existing Oncotree code
```

Comment:
minor point: Originally pandas used NaN to represent missing data. This has resulted in a whole lot of headaches and I think the authors and community now have accepted that was a bad idea. To remedy this, at some point pandas introduced new data types which are aware of a special NA value. 

When writing new code, I recommend using pandas's NA instead of NaN because it can avoid confusion/problems. In this context, it just is a matter of changing this line to

```
new_node['OncotreeCode'] = pd.na
```